### PR TITLE
Add Arabic localization

### LINKED ISSUE #1183: Add Arabic language localisation
Hi. I saw that your library doesn't support Arabic localisation. As a native speaker, I'd like to contribute to add support for it. Thanks.

### PR DESCRIPTION
**Pull request by abh31000**
## Proposed change

Added localization for Arabic-language countries: Algeria, Bahrain, Egypt, Morocco, Saudi Arabia, Tunisia, United Arab Emirates (ar, EN_US locales).
I don't know if this PR should close #1183 yet.

Notes:
- Some countries' holidays need an update, I will work on them on separate PRs
- Morocco's holidays were originally in french
- I noticed some parsing issues in Arabic when multiple holidays fall on the same day
- UAE's estimated holidays don't display the default or custom estimated label

PS I apologize for the extended time it took to complete this PR.

## Type of change

- [ ] New country/market holidays support (thank you!)
- [x] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency upgrade (version update)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] This PR is filed against `beta` branch of the repository
- [x] This PR doesn't contain any merge conflicts and has clean commit history
- [x] The code style looks good: `make pre-commit`
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)
- [x] The related [documentation][docs] has been added/updated (check off the box for free if no updates is required)

[contributing-guidelines]: https://github.com/dr-prodigy/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/dr-prodigy/python-holidays/tree/beta/docs/source

### CODE REVIEW

**Review by KJhellico — COMMENTED** [review]
It's a great work! 👍 Please look at the remarks below.

File: holidays/locale/ar/LC_MESSAGES/AE.po [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# United Arab Emirates holidays ar localization.
```
Comment (arkid15r) [inline holidays/locale/ar/LC_MESSAGES/AE.po:1]: For ar/LC_MESSAGES/*.po it's just holidays (w/o localization)
```suggestion
# United Arab Emirates holidays.
```

File: holidays/locale/ar/LC_MESSAGES/AE.po [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+# Authors: Abdelkhalek Boukli Hacene <bouklihabd31@gmail.com>, 2023.
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Python Holidays 0.28\n"
+"POT-Creation-Date: 2023-06-28 00:13+0100\n"
+"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
```
Comment (arkid15r) [inline holidays/locale/ar/LC_MESSAGES/AE.po:8]: Some metadata is missing, perhaps this was the very first item translated.

File: holidays/countries/united_arab_emirates.py [hunk 3]
```diff
@@ -39,6 +41,9 @@ class UnitedArabEmirates(HolidayBase, InternationalHolidays, IslamicHolidays):
     """
 
     country = "AE"
+    default_language = "ar"
+    estimated_label = tr("(تقدير*) *%s")
```
Comment (arkid15r) [inline holidays/countries/united_arab_emirates.py:45]:
```suggestion
    # Estimated label.
    estimated_label = tr("(تقدير*) *%s")
```

File: holidays/countries/united_arab_emirates.py [hunk 4]
```diff
@@ -50,40 +55,38 @@ def _populate(self, year):
         super()._populate(year)
 
         # New Year's Day.
-        self._add_new_years_day("New Year's Day")
+        self._add_new_years_day(tr("رأس السنة الميلادية"))
 
         # Commemoration Day.
         if year >= 2015:
-            self._add_holiday("Commemoration Day", *((DEC, 1) if year >= 2019 else (NOV, 30)))
+            self._add_holiday(tr("يوم الشهيد"), *((DEC, 1) if year >= 2019 else (NOV, 30)))
```
Comment (arkid15r) [inline holidays/countries/united_arab_emirates.py:62]:
```suggestion
            # Commemoration Day.
            self._add_holiday(tr("يوم الشهيد"), *((DEC, 1) if year >= 2019 else (NOV, 30)))
```

File: holidays/countries/united_arab_emirates.py [hunk 5]
```diff
@@ -50,40 +55,38 @@ def _populate(self, year):
...
 
         # Arafat Day & Eid al-Adha.
-        self._add_arafah_day("Arafat (Hajj) Day")
+        self._add_arafah_day(tr("وقفة عرفة"))
 
-        eid_al_adha = "Eid al-Adha"
-        self._add_eid_al_adha_day(eid_al_adha)
-        self._add_eid_al_adha_day_two(f"{eid_al_adha} Holiday")
-        self._add_eid_al_adha_day_three(f"{eid_al_adha} Holiday")
+        self._add_eid_al_adha_day(tr("عيد الأضحى"))
```
Comment (arkid15r) [inline holidays/countries/united_arab_emirates.py:76]:
```suggestion
        # Eid al-Adha.
        self._add_eid_al_adha_day(tr("عيد الأضحى"))
```

Could you also add a comment for the 2 following lines?

File: holidays/countries/bahrain.py [hunk 6]
```diff
@@ -25,6 +27,9 @@ class Bahrain(HolidayBase, InternationalHolidays, IslamicHolidays):
     """
 
     country = "BH"
+    default_language = "ar"
+    estimated_label = tr("(تقدير*) *%s")
```
Comment (arkid15r) [inline holidays/countries/bahrain.py:31]:
```suggestion
    # Estimated label.
    estimated_label = tr("(تقدير*) *%s")
```

File: holidays/countries/bahrain.py [hunk 7]
```diff
@@ -36,37 +41,37 @@ def _populate(self, year):
...
         # Al Hijra New Year.
-        hijri_new_year = "Al Hijra New Year"
+        hijri_new_year = tr("رأس السنة الهجرية")
         self._add_islamic_new_year_day(hijri_new_year)
 
         # Ashura.
-        ashura = "Ashura"
-        self._add_ashura_eve(f"{ashura} Holiday")
+        ashura = tr("عاشورة")
+        self._add_ashura_eve(tr("ليلة عاشورة"))
```
Comment (arkid15r) [inline holidays/countries/bahrain.py:67]:
```suggestion
        # Ashura eve.
        self._add_ashura_eve(tr("ليلة عاشورة"))
```

File: holidays/countries/saudi_arabia.py [hunk 8]
```diff
@@ -49,17 +54,21 @@ def _add_islamic_observed(self, hol_name: str, hol_dates: Set[date]) -> None:
         for dt in hol_dates:
             weekend_days = sum(self._is_weekend(dt + td(days=i)) for i in range(4))
             for i in range(weekend_days):
-                self._add_holiday(f"{hol_name} (Observed)", dt + td(days=+4 + i))
+                self._add_holiday(self.tr("(ملاحظة) %s") % self.tr(hol_name), dt + td(days=+4 + i))
```
[the remark this answers is not in the document]
Comment (abh31000) [reply holidays/countries/saudi_arabia.py:57]: Arabic is not a gender neutral language. So the word may change slightly.
And I think the translation in Canada is just slightly wrong. "تمت ملاحظته" translates to "Has been observed".

File: holidays/countries/egypt.py [hunk 9]
```diff
@@ -42,54 +47,54 @@ def _populate(self, year):
         super()._populate(year)
 
         # New Year's Day
-        self._add_new_years_day("New Year's Day - Bank Holiday")
+        self._add_new_years_day(tr("رأس السنة الميلادية - عطلة بنوك"))
```
Comment (arkid15r) [inline holidays/countries/egypt.py:50]: I'm not sure if it makes sense  to keep "Bank Holiday" here unless it's an official holiday name.

File: holidays/countries/egypt.py [hunk 10]
```diff
@@ -42,54 +47,54 @@ def _populate(self, year):
...
 
         # Coptic Christmas
-        self._add_christmas_day("Coptic Christmas")
+        self._add_christmas_day(tr("عيد الميلاد المجيد (تقويم قبطي)"))
 
         # 25th of Jan
         if year >= 2012:
-            self._add_holiday("Revolution Day - January 25", JAN, 25)
+            self._add_holiday(tr("عيد ثورة 25 يناير"), JAN, 25)
```
Comment (arkid15r) [inline holidays/countries/egypt.py:57]: As well as "January 25" unless it's a part of the official holiday name.

**Review by arkid15r — COMMENTED** [review]
That's a really great l10n work! Thanks for adding this as promised!

I've gone through AE and BH .po files and added suggestions for missing translator comments. It'd be really useful to have them for all holiday names in the localized countries as those are no longer English language names.

Again, thanks you for this great contribution!

File: holidays/countries/egypt.py (hunk 9 above)
Comment (arkid15r) [reply holidays/countries/egypt.py:51]: It'd be great to address it in a separate PR. For now you can keep it as is or leave a TODO if you'd like.

File: holidays/countries/egypt.py (hunk 10 above)
Comment (abh31000) [reply holidays/countries/egypt.py:59]: It's part of the official holiday name I believe.
[Reference](https://www.presidency.eg/en/%D9%85%D8%B5%D8%B1/%D8%A7%D9%84%D8%B9%D8%B7%D9%84%D8%A7%D8%AA-%D8%A7%D9%84%D8%B1%D8%B3%D9%85%D9%8A%D8%A9/)

File: holidays/countries/morocco.py [hunk 11]
```diff
@@ -42,72 +49,71 @@ def _populate(self, year):
...
 
-        # Youth day
         if year >= 2001:
-            self._add_holiday("Fête de la jeunesse", AUG, 21)
+            dt = (AUG, 21)
         else:
-            self._add_holiday("Fête du Trône", JUL, 9)
+            dt = (JUL, 9)
+        # Youth day
+        self._add_holiday(tr("عيد الشباب"), *dt)
```
Comment (arkid15r) [inline holidays/countries/morocco.py:87]: self._add_holiday(tr("عيد الشباب"), *((AUG, 21) if year >= 2001 else (JUL, 9)))

**Review by arkid15r — COMMENTED** [review]
Thank you for addressing most of the comments!
Please look into some minor issues I spotted during another review round.

I'd also like to make sure that your initial concerns are resolved:

> Notes:
>   - Some countries' holidays need an update, I will work on them on separate PRs
>   - Morocco's holidays were originally in french
>   - I noticed some parsing issues in Arabic when multiple holidays fall on the same day
>   - UAE's estimated holidays don't display the default or custom estimated label

Should we look into any of these specifically?

Thank you!

**Comment by KJhellico** [comment]
> When multiple holidays fall on the same day, In Arabic the words get all scrambled.

I tried `DZ.get_list("2025-07-05")` (date with multiple holidays in Algeria) and everything looks good. What exactly is wrong?

> When I was writing l10n tests for UAE, I encountered some errors. Both the `en_US` and `ar` holidays were missing the default and custom estimated labels. What is it that could cause this issue ? Thank you.

UAE has exact dates of Islamic holidays for 2017-2019 (class `UnitedArabEmiratesIslamicCalendar`), so estimated label is not displayed in these years.

**Comment by abh31000** [comment]
> I tried `DZ.get_list("2025-07-05")` (date with multiple holidays in Algeria) and everything looks good. What exactly is wrong?

Yes you're right. The parsing only looked wrong when assertion errors were raised when I wrote testcases for dates with multiple holidays.

**Comment by KJhellico** [comment]
It has been merged into beta. @abh31000 , thank you for this great contribution!

**Pull request merged**